src/shared/dissect-image.c

```diff
@@ -1409,56 +1417,89 @@ int decrypted_image_relinquish(DecryptedImage *d) {
-                e = endswith(fn, ".raw");
+                verity_filename = new(char, strlen(image) + STRLEN(".verity") + 1);
+                if (!verity_filename)
+                        return -ENOMEM;
+                strcpy(verity_filename, image);
+                e = endswith(verity_filename, ".raw");
                 if (e)
-                        n = e;
-
-                strcpy(n, ".roothash");
+                        strcpy(e, ".verity");
+                else
+                        strcat(verity_filename, ".verity");
 
-                r = read_one_line_file(fn, &text);
-                if (r == -ENOENT) {
-                        *ret = NULL;
-                        *ret_size = 0;
-                        return 0;
+                r = access(verity_filename, R_OK);
+                if (r < 0) {
+                        if (errno != ENOENT)
+                                return -errno;
+                        free(verity_filename);
+                        verity_filename = NULL;
                 }
-                if (r < 0)
-                        return r;
         }
 
-        r = unhexmem(text, strlen(text), &k, &l);
-        if (r < 0)
-                return r;
-        if (l < sizeof(sd_id128_t))
-                return -EINVAL;
+        if (ret_roothash) {
+                _cleanup_free_ char *text = NULL;
+                assert(ret_roothash_size);
 
-        *ret = TAKE_PTR(k);
-        *ret_size = l;
+                r = getxattr_malloc(image, "user.verity.roothash", &text, true);
+                if (r < 0) {
+                        char *fn, *e, *n;
+
+                        if (!IN_SET(r, -ENODATA, -EOPNOTSUPP, -ENOENT))
+                                return r;
+
+                        fn = newa(char, strlen(image) + STRLEN(".roothash") + 1);
+                        n = stpcpy(fn, image);
+                        e = endswith(fn, ".raw");
+                        if (e)
+                                n = e;
+
+                        strcpy(n, ".roothash");
+
+                        r = read_one_line_file(fn, &text);
+                        if (r == -ENOENT) {
+                                *ret_roothash = NULL;
+                                *ret_roothash_size = 0;
+                        } else if (r < 0)
+                                return r;
+                }
+
+                if (text) {
+                        r = unhexmem(text, strlen(text), &roothash_decoded, &roothash_decoded_size);
+                        if (r < 0)
+                                return r;
+                        if (roothash_decoded_size < sizeof(sd_id128_t))
+                                return -EINVAL;
+                }
+        }
+
+        if (roothash_decoded) {
+                *ret_roothash = TAKE_PTR(roothash_decoded);
+                *ret_roothash_size = roothash_decoded_size;
+        }
+        if (verity_filename)
+                *ret_verity_data = TAKE_PTR(verity_filename);
```

Comment:
as mentioned: our rule is to unconditionally initialize returm values on success. 

this, these if checks should just not be done, but the three asignments should be done here unconditionally. (you can then drop the early reset to NULL/0 further up too) The code becomes a lot simpler, and the return params are initialized at once place only.

Reply:
Ok - I've removed the earlier reset, and changed the check to look at the return buffers instead.

It is possible for one return buffer to be passed and one to be null, as it's legal for example to call nspawn --root-hash (thus this function gets a null as we don't want to overwrite what the user set explicitly) and let the auto-discovery find the verity file, or viceversa